torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -287,30 +300,28 @@ struct ChunkDatasetOptions {
 /// inspired by this paper http://martin.zinkevich.org/publications/nips2010.pdf
 template <
     typename ChunkReader,
-    typename ChunkSampler = samplers::RandomSampler,
     typename ExampleSampler = samplers::RandomSampler>
-class ChunkDataset final
-    : public StatefulDataset<
-          ChunkDataset<ChunkReader, ChunkSampler, ExampleSampler>,
-          typename ChunkReader::BatchType,
-          size_t> {
+class ChunkDataset final : public StatefulDataset<
+                               ChunkDataset<ChunkReader, ExampleSampler>,
+                               typename ChunkReader::BatchType,
+                               size_t> {
  public:
   using BatchType = torch::optional<typename ChunkReader::BatchType>;
   using UnwrappedBatchType = typename ChunkReader::BatchType;
   using BatchRequestType = size_t;
-  using ChunkSamplerType = ChunkSampler;
   using ExampleSamplerType = ExampleSampler;
 
   ChunkDataset(
       ChunkReader chunk_reader,
-      ChunkSampler chunk_sampler,
       ExampleSampler example_sampler,
+      std::shared_ptr<samplers::DistributedSampler> chunk_sampler,
```

Comment:
Shall this be an unique_ptr? If it is a shared_ptr, the user may change the sampler state outside ChunkDataset, which is undesired. With an unique_ptr, after it is moved inside ChunkDataset, the object is uniquely owned by ChunkDataset. What do you think?

Reply:
The user needs to call the set_epoch() on the sampler and so need to have a reference. There are several alternatives to this approach. We can move the set_epoch() to StatefulDataloader, or automatically increment the epoch value with the iteration. I followed the python API in pytorch.